### ISSUE TITLE
Question: Reloading a large AOF file (and running out of memory)

### ISSUE THREAD

**Issue by thisisaaronland** [tidwall/tile38#71]
This is a by-product of issue #70 
1. We have an AOF file (produced by a version 1.5.1 instance of Tile38) that is ~17GB on disk and, when its all said and done, consumes about 24GB RAM on a 32GB machine. This is an AOF that has not been `SHRINK` -ed (so maybe that's the root problem?)
2. After backporting to version 1.4.2 we spun up Tile38 and watched it import the AOF file eventually running out of memory. 
3. After upgrading to version 1.5.2 we spun up Tile38 and watched it import the AOF file eventually running out of memory, again.

Is this a known-known and if so are there any best practices for mitigating against running out of memory when starting up a Tile38 server? Is there a relationship between the size of the AOF file on disk and memory requirements for both startup and general operations?

(As I write this simply re-feeding the index is not a big deal.)

Cheers,

![screen shot 2016-10-20 at 18 29 31](https://cloud.githubusercontent.com/assets/12658759/19583812/93885ae8-96f4-11e6-9cf4-00ac11110c80.png)

**Comment by thisisaaronland** [comment]
I guess the follow-on question is: Is there a way to `SHRINK` the AOF file prior to or during import?

**Comment by tidwall** [comment]
24GB of mem for a 17GB AOF seems like a lot. Does Tile38 crash when it runs out of memory? I've yet to experience an oom crash. If so, is this happening on a system with virtual memory? 

At this time, there isn't a way to shrink the AOF until all of the data is loaded into memory. Perhaps a commandline option at startup.

I'm going to need to investigate further.

**Comment by thisisaaronland** [comment]
It's just a plain vanilla EC2/Ubuntu machine with defaults. I can reproduce the problem pretty reliably (see attached). In the meantime I will try throwing hardware at the problem and see what happens.

![screen shot 2016-10-21 at 11 57 57](https://cloud.githubusercontent.com/assets/12658759/19608644/7b029270-9787-11e6-91bf-f5a2a1c460f5.png)

**Comment by thisisaaronland** [comment]
Passing these along in case they are helpful:

![screen shot 2016-10-21 at 13 10 26](https://cloud.githubusercontent.com/assets/12658759/19610260/5c3432e6-9790-11e6-8f77-2f873980772f.png)
![screen shot 2016-10-21 at 13 13 27](https://cloud.githubusercontent.com/assets/12658759/19610259/5c33ff42-9790-11e6-92cc-68d1dd5f477a.png)

**Comment by thisisaaronland** [comment]
Also, this:

```
127.0.0.1:9851> server
{"ok":true,"stats":{"aof_size":17262573355,"avg_item_size":85,"heap_size":21084301592,"id":"2f275de4e8c4389d6f5aa023fbdf7f7c","in_memory_size":9021741616,"max_heap_size":0,"num_collections":1,"num_hooks":0,"num_objects":52039408,"num_points":245603205,"pointer_size":8,"read_only":false},"elapsed":"59.717278ms"}
127.0.0.1:9851> AOFSHRINK
{"ok":true,"elapsed":"228ns"}
```

At which point the heap size starts to go a little crazy...

```
127.0.0.1:9851> server
{"ok":true,"stats":{"aof_size":17262573355,"avg_item_size":101,"heap_size":24917315104,"id":"2f275de4e8c4389d6f5aa023fbdf7f7c","in_memory_size":9021741616,"max_heap_size":0,"num_collections":1,"num_hooks":0,"num_objects":52039408,"num_points":245603205,"pointer_size":8,"read_only":false},"elapsed":"59.266378ms"}
127.0.0.1:9851> server
{"ok":true,"stats":{"aof_size":17262573355,"avg_item_size":131,"heap_size":32398004840,"id":"2f275de4e8c4389d6f5aa023fbdf7f7c","in_memory_size":9021741616,"max_heap_size":0,"num_collections":1,"num_hooks":0,"num_objects":52039408,"num_points":245603205,"pointer_size":8,"read_only":false},"elapsed":"86.35419ms"}
127.0.0.1:9851> server
{"ok":true,"stats":{"aof_size":17262573355,"avg_item_size":143,"heap_size":35366405944,"id":"2f275de4e8c4389d6f5aa023fbdf7f7c","in_memory_size":9021741616,"max_heap_size":0,"num_collections":1,"num_hooks":0,"num_objects":52039408,"num_points":245603205,"pointer_size":8,"read_only":false},"elapsed":"82.671635ms"}
```

![screen shot 2016-10-21 at 13 18 48](https://cloud.githubusercontent.com/assets/12658759/19610364/10ac166c-9791-11e6-886e-a353b2c6eb6b.png)

**Comment by tidwall** [comment]
Interesting. Does the heap lower after a while? Have you tried the GC command. (Executes a garbage collection)?

**Comment by thisisaaronland** [comment]
The heap size does come down, eventually:

```
server
{"ok":true,"stats":{"aof_size":16533371938,"avg_item_size":85,"heap_size":21084283600,"id":"2f275de4e8c4389d6f5aa023fbdf7f7c","in_memory_size":9021741616,"max_heap_size":0,"num_collections":1,"num_hooks":0,"num_objects":52039408,"num_points":245603205,"pointer_size":8,"read_only":false},"elapsed":"76.364715ms"}
```

_Suddenly notices there is a `GC` command..._

**Comment by thisisaaronland** [comment]
Executing `GC` returns:

```
127.0.0.1:9851> GC
{"ok":true,"elapsed":"377ns"}
```

All 8 CPUs spiked briefly and then went back down to hovering around 0. There is no significant change in memory usage. `htop` reports 50.1G total (VIRT) and 30.2GB resident (RES).

**Comment by thisisaaronland** [comment]
FYI, this is what the graphs look like stopping and restarting the server (twice more) with the same AOF file after all the steps above:

![screen shot 2016-10-21 at 14 42 09](https://cloud.githubusercontent.com/assets/12658759/19613028/e1892f10-979e-11e6-800a-c9e36faee60d.png)
![screen shot 2016-10-21 at 14 57 33](https://cloud.githubusercontent.com/assets/12658759/19613029/e18e00c6-979e-11e6-9e0d-6281d08d59e5.png)

**Comment by tidwall** [comment]
Thanks for the all the data. 

I'm looking to find ways to lower the memory usage on large datasets like the one you are working with. I have a few ideas on how to do so, but I'd like to produce a dataset similar to yours. 

The SERVER command you provided gave a lot of details to the number of objects, point / object ratio, heap size, etc.

But could you help with a couple more questions:

1. What's the average size of object IDs?  You showed one ID as "85922583#whosonfirst-data", is this a typical length?
2. How often do you use just Points as opposed to Polygons? 
3. Do you have any 3D objects using the Z coordinate?

I made some big changes between 1.4 and 1.5 with the R-tree structure, added 3D support, and removed the Quadtree all together. Maybe one of these is causing an increase in allocations of pointers.

I'll keep you posted.

**Comment by thisisaaronland** [comment]
The _average_ size is probably pretty small when you factor in the fact that most of the records will be venues, meaning points or in rare occasions simple polygons. No 3D objects.

That said there are about 4M simple to very complex polygons for postal codes and administrative places:

https://whosonfirst.mapzen.com/spelunker/placetypes/

One of the tenets of Who's On First is that any given record [can have multiple geometries](https://mapzen.com/blog/mapping-with-bias/#geometries), including simplified or display geometries. That said right now they don't (have simplified geometries) and some national boundaries are enormous and fiddly – like New Zealand:

https://whosonfirst.mapzen.com/data/856/333/45/85633345.geojson

So _most_ records (20M+) will be very small but some of them will be very large, relative to all the others. Ideally we'd like to be able to index all the places (24M and growing) in a single index rather than breaking things out by placetype or some other grouping.

If you're looking for test data you can grab stuff here:

https://whosonfirst.mapzen.com/bundles/

Venue bundled are also available just not listed on the page, because venues are hard ([the long version is over here](https://mapzen.com/blog/whosonfirst-venues/)). For example:

https://whosonfirst.mapzen.com/bundles/wof-venue-us-ca-latest-bundle.tar.bz2

All of which should be able to be imported in to Tile38 using this:

https://github.com/whosonfirst/go-whosonfirst-tile38/blob/master/README.md#example

**Comment by tidwall** [comment]
Your Who's On First project is really cool. I'm downloading and testing data from the links you provided. I'm hoping to get some insights over the next couple days. Thanks.

**Comment by tidwall** [comment]
I just pushed an update to a new [memoptz](https://github.com/tidwall/tile38/tree/memoptz) branch. It's only a start but I made some simple changes that brought the heap size down about 5-10%. 

Most of the extra overhead is coming from interfaces and pointer allocations in Go. I'm looking for ways to minimize as much as possible.

**Comment by thisisaaronland** [comment]
Would it be helpful for me to throw our current AOF file at that branch or is it too soon?

**Comment by tidwall** [comment]
Sure, if it's not too much trouble or time. Not major hurry from my side though.

**Comment by thisisaaronland** [comment]
Finally got a few spare cycles to try this with the `memoptz` branch. The span of the graphs below is about 30 minutes.

```
Starting tile38-server in debug mode...
sudo -u tile38 /usr/local/bin/tile38-server -d /usr/local/var/tile38

   _______ _______
  |       |       |
  |____   |   _   |   Tile38 1.5.2 (36ab60b) 64 bit (amd64/linux)
  |       |       |   Port: 9851, PID: 7717
  |____   |   _   |
  |       |       |   tile38.com
  |_______|_______|

2016/10/26 23:35:48 [INFO] Server started, Tile38 version 1.5.2, git 36ab60b
```

![screen shot 2016-10-26 at 17 06 44](https://cloud.githubusercontent.com/assets/12658759/19749559/c6f8a906-9b9e-11e6-91d0-2855b97480a8.png)

**Comment by umpc** [comment]
@tidwall This may or may not help, but removing the interface-typed ID field and replacing the Properties ```map[string]interface{}``` with a ```[]byte``` containing marshaled JSON, in the ```Feature``` type, reduced HeapAlloc by roughly 30% in my case (all Objects are Features).

This was using master, so the memoptz branch could reduce it further.

At the risk of stating the obvious:
```
type Feature struct {
	Geometry   Object
	BBox       *BBox
	Properties []byte
}

func fillFeatureMap(m map[string]interface{}) (Feature, []byte, error) {
	...
	case map[string]interface{}:
		g.Properties, err = json.Marshal(v)
	}
	return g, nil, err
}

func (g Feature) JSON() string {
	...
	if g.Properties != nil {
		buf.WriteString(`,"properties":`)
		buf.Write(g.Properties)
	}
	buf.WriteByte('}')
	return buf.String()
}
```
etc.

**Comment by tidwall** [comment]
@umpc Great tip. I went ahead and removed the `map[string]interface{}` and added a string field that contains the id+properties. This should be very small in memory without losing the Id.

I didn't actually see an improvment in HeapAlloc from the test bundle that @thisisaaronland provided, though I don't think features were being used in that dataset.

I also removed the "encoding/json" package in favor of "tidwall/gjson" which sped up the parsing by about 25%.

These changed are posted to the memoptz branch.

**Comment by thisisaaronland** [comment]
I will pull from `memoptz` and try a re-index later today.

**Comment by thisisaaronland** [comment]
Just to confirm, those changes to the `memoptz` branch help. As in: the Tile38 server is able to load the AOF file and serve requests:

```
2016/11/02 21:08:31 [INFO] Server started, Tile38 version 1.5.2, git bdcbc9c
2016/11/02 21:33:49 [INFO] AOF loaded 52039408 commands: 1518.20s, 34277/s, 11 MB/s
```
Updated memoptz branch:

![screen shot 2016-11-02 at 14 47 28](https://cloud.githubusercontent.com/assets/12658759/19948821/db7b7c56-a10b-11e6-8009-94cc0c52a5df.png)

Old and new memoptz branches (with the old one getting oom-killed):

![screen shot 2016-11-02 at 14 47 48](https://cloud.githubusercontent.com/assets/12658759/19948820/db7aaace-a10b-11e6-9050-f14bdd061845.png)

**Comment by tidwall** [comment]
@thisisaaronland Moving in the right direction. I think the changes @umpc recommended cut down on the pointer allocations quite a bit. 

Also I haven't had a chance to test the Go 1.8 beta yet, but it's suppose to provide some big improvements with the memory management arena. I hoping to get around to it in the next few day.

**Comment by umpc** [comment]
You might not be aware of this, but it appears that the loop in the new ```Feature.getRaw()``` method, used to split the ID and Properties, takes a significant amount of time, depending on the workload.

```
20.01s  1.21% 64.25%    813.83s 49.20%  github.com/tidwall/tile38/index/rtree.d3search
19.33s  1.17% 65.41%     19.33s  1.17%  runtime.futex
17.35s  1.05% 66.46%     17.35s  1.05%  encoding/json.stateInString
16.60s  1.00% 67.47%     16.60s  1.00%  sync/atomic.AddUint32
16.33s  0.99% 68.45%     16.33s  0.99%  encoding/base64.(*Encoding).Encode
15.18s  0.92% 69.37%     21.41s  1.29%  time.nextStdChunk
14.87s   0.9% 70.27%     20.54s  1.24%  runtime.(*mheap).freeSpanLocked
14.03s  0.85% 71.12%     14.03s  0.85%  github.com/tidwall/tile38/geojson.Feature.getRaw
```

Seeing how the ID is already stored separately, is always passed into the iterator function for each result, and no code appears to depend on having the ID included in the marshaled JSON specifically, I am curious as to why it's worth the extra time and memory?

I can understand why bbox info from the pointer might be included in the marshaled output, but not the ID itself, because the ID is always being provided by the iterator anyway.

Thoughts?

**Comment by tidwall** [comment]
I just pushed an update that should help with performance of the `getRaw` function.

Tile38 only cares about the collection ID:
```
> SET trucks 89234 OBJECT {"type":"Feature","id":89234,"properties":{...},"geometry":{...}}
             ^^^^^                                             
```
The GeoJSON "id" field doesn't really matter to Tile38 and it can be any valid [json object](http://geojson.org/geojson-spec.html#feature-objects), even null or non-existent:
```
> SET trucks 89234 OBJECT {"type":"Feature","id":89234,"properties":{...},"geometry":{...}}
                                                 ^^^^^
```

The same applies to the json "properties" field. Neither field is used by Tile38. 

> I am curious as to why it's worth the extra time and memory?

Only for completeness to the GeoJSON specification. I figured that it's a good idea to return all fields that are used in the spec.

One possible solution: Because both "id" and "properties" are not required by Tile38 perhaps I could add a special keyword to the `SET` command that would tell the collection to ignore GeoJSON fields that are not needed by the spatial index.  

```
> SET trucks 89234 TRUNC OBJECT {"type":"Feature","id":89234,"properties":{...},"geometry":{...}}
> GET trucks 89234
{"type":"Feature","geometry":{...}}
```

This could potentially save lots of memory depending on the size of the id and number of properties. The downside is that "id" and "properties" would be completely lost.

Thought?

**Comment by umpc** [comment]
> Only for completeness to the GeoJSON specification. I figured that it's a good idea to return all fields that are used in the spec.

You're right. I had overlooked those from the spec and only had "type," "geometry," and "properties" (which are what I use) in mind.

With Tile38 being used for geofencing and similar actions that may not have properties, I could see something like this being justified, though in my opinion, I think that this kind of truncation would be best left to the user/developer in order to keep the redis and package APIs lean.

In other words, since "id" is part of the spec, I think that the time and memory can be justified.

If I come up with a more efficient way of handling this, (I've not checked out your fix yet) I'll reply here or submit a PR. Thank you for looking into this.

**Comment by tidwall** [comment]
Good point, I agree that we should just keep the API as is.

**Comment by thisisaaronland** [comment]
Passing along FYI. This is what I see when I spin up the (17GB) AOF file with the latest version of the `memoptz` branch:

```
 /usr/local/bin/tile38-server -d /usr/local/var/tile38

   _______ _______
  |       |       |
  |____   |   _   |   Tile38 1.5.2 (1ac6ad9) 64 bit (amd64/linux)
  |       |       |   Port: 9851, PID: 7899
  |____   |   _   |
  |       |       |   tile38.com
  |_______|_______|

2016/11/09 17:34:44 [INFO] Server started, Tile38 version 1.5.2, git 1ac6ad9
2016/11/09 18:00:39 [INFO] AOF loaded 52039408 commands: 1555.05s, 33465/s, 11 MB/s
2016/11/09 18:00:39 [INFO] The server is now ready to accept connections on port 9851

127.0.0.1:9851> server
{"ok":true,"stats":{"aof_size":17262573355,"avg_item_size":73,"heap_size":18100647272,"id":"683dd04bfb1c4756d487f31c6c67cf43","in_memory_size":9021741616,"max_heap_size":0,"mem_alloc":18100647272,"num_collections":1,"num_hooks":0,"num_objects":52039408,"num_points":245603205,"num_strings":26019704,"pointer_size":8,"read_only":false},"elapsed":"98.804498ms"}
127.0.0.1:9851> GC
{"ok":true,"elapsed":"1.137µs"}
127.0.0.1:9851> server
{"ok":true,"stats":{"aof_size":17262573355,"avg_item_size":73,"heap_size":18100582064,"id":"683dd04bfb1c4756d487f31c6c67cf43","in_memory_size":9021741616,"max_heap_size":0,"mem_alloc":18100582064,"num_collections":1,"num_hooks":0,"num_objects":52039408,"num_points":245603205,"num_strings":26019704,"pointer_size":8,"read_only":false},"elapsed":"54.609214ms"}
```

![screen shot 2016-11-09 at 10 14 39](https://cloud.githubusercontent.com/assets/12658759/20149266/60a8691c-a665-11e6-8114-d762e98a48c2.png)

**Comment by tidwall** [comment]
FYI, I merged the memoptz branch with master yesterday.

**Comment by thisisaaronland** [comment]
Thanks!

We managed to again blow up our memory usage so we've shelved that particular use of Tile38 but I think that we will continue to use it (with less data overall) in a different setting.

I will try to test all of this out shortly.

**Comment by tidwall** [comment]
You're welcome and thanks a bunch for helping work through the optimizations.

**Comment by m1ome** [comment]
@tidwall do we still need this one? It looks like all stuff are done to help memory opts?

**Comment by Lars-Meijer** [comment]
I might be mistaken but, after a flushdb command, the AOF can be thrown away right (or maybe renamed to appendonly.bak)? 
Its not really that important but maybe just a quality of life improvement if someone is gonna work on this.

**Comment by tidwall** [comment]
@m1ome I agree that there's not much more we can do as it stands today.

@Lars-Meijer Good idea. A FLUSHDB should simply shrink the the AOF to an empty file or trigger an autoshrink in the background #74.

**Comment by tidwall** [comment]
I’m closing this issue because it has been inactive for many months. 

Please feel free to reopen if needed.

Thank you!
